diff --git a/.cargo/config.toml b/.cargo/config.toml
index e11d56a31..c9c774bc1 100644
--- a/.cargo/config.toml
+++ b/.cargo/config.toml
@@ -1,2 +1,6 @@
 [alias]
 xtask = "run --release --package xtask --"
+
+[target.x86_64-unknown-linux-gnu]
+rustflags = ["-Cforce-frame-pointers=yes", "-Cstrip=none"]
+#rustflags = ["-Cforce-frame-pointers=yes", "-Cstrip=none", "-Zinline-mir=no"]
diff --git a/Cargo.toml b/Cargo.toml
index e4efedab9..fc955a7e2 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -40,8 +40,8 @@ codegen-units = 1
 # We now compile heed without the NDEBUG define for better performance.
 # However, we still enable debug assertions for a better detection of
 # disk corruption on the cloud or in OSS.
-[profile.release.package.heed]
-debug-assertions = true
+#[profile.release.package.heed]
+#debug-assertions = true
 
 [profile.dev.package.flate2]
 opt-level = 3
@@ -50,3 +50,6 @@ opt-level = 3
 opt-level = 3
 [profile.dev.package.roaring]
 opt-level = 3
+
+[patch.crates-io]
+roaring = { path = "/home/yugr/tasks/rust/bench/mod/roaring-0.10.12" }
diff --git a/crates/benchmarks/Cargo.toml b/crates/benchmarks/Cargo.toml
index f05100c2c..760254a40 100644
--- a/crates/benchmarks/Cargo.toml
+++ b/crates/benchmarks/Cargo.toml
@@ -39,23 +39,3 @@ default = ["milli/all-tokenizations"]
 [[bench]]
 name = "search_songs"
 harness = false
-
-[[bench]]
-name = "search_wiki"
-harness = false
-
-[[bench]]
-name = "search_geo"
-harness = false
-
-[[bench]]
-name = "indexing"
-harness = false
-
-[[bench]]
-name = "sort"
-harness = false
-
-[[bench]]
-name = "filter_starts_with"
-harness = false
diff --git a/crates/benchmarks/benches/search_songs.rs b/crates/benchmarks/benches/search_songs.rs
index e1cbb5730..c5fc11a3b 100644
--- a/crates/benchmarks/benches/search_songs.rs
+++ b/crates/benchmarks/benches/search_songs.rs
@@ -32,16 +32,7 @@ fn base_conf(builder: &mut Settings) {
 const BASE_CONF: Conf = Conf {
     dataset: datasets_paths::SMOL_SONGS,
     queries: &[
-        "john ",             // 9097
-        "david ",            // 4794
-        "charles ",          // 1957
-        "david bowie ",      // 1200
-        "michael jackson ",  // 600
-        "thelonious monk ",  // 303
-        "charles mingus ",   // 142
-        "marcus miller ",    // 60
         "tamo ",             // 13
-        "Notstandskomitee ", // 4
     ],
     configure: base_conf,
     primary_key: Some("id"),
@@ -69,126 +60,12 @@ fn bench_songs(c: &mut criterion::Criterion) {
 
     #[rustfmt::skip]
     let confs = &[
-        /* first we bench each criterion alone */
-        utils::Conf {
-            group_name: "proximity",
-            queries: &[
-                "black saint sinner lady ",
-                "les dangeureuses 1960 ",
-                "The Disneyland Sing-Along Chorus ",
-                "Under Great Northern Lights ",
-                "7000 Danses Un Jour Dans Notre Vie ",
-            ],
-            criterion: Some(&["proximity"]),
-            optional_words: false,
-            ..BASE_CONF
-        },
-        utils::Conf {
-            group_name: "typo",
-            queries: &[
-                "mongus ",
-                "thelonius monk ",
-                "Disnaylande ",
-                "the white striper ",
-                "indochie ",
-                "indochien ",
-                "klub des loopers ",
-                "fear of the duck ",
-                "michel depech ",
-                "stromal ",
-                "dire straights ",
-                "Arethla Franklin ",
-            ],
-            criterion: Some(&["typo"]),
-            optional_words: false,
-            ..BASE_CONF
-        },
-        utils::Conf {
-            group_name: "words",
-            queries: &[
-                "the black saint and the sinner lady and the good doggo ", // four words to pop
-                "les liaisons dangeureuses 1793 ",                         // one word to pop
-                "The Disneyland Children's Sing-Alone song ",              // two words to pop
-                "seven nation mummy ",                                     // one word to pop
-                "7000 Danses / Le Baiser / je me trompe de mots ",         // four words to pop
-                "Bring Your Daughter To The Slaughter but now this is not part of the title ", // nine words to pop
-                "whathavenotnsuchforth and a good amount of words to pop to match the first one ", // 13
-            ],
-            criterion: Some(&["words"]),
-            ..BASE_CONF
-        },
-        utils::Conf {
-            group_name: "asc",
-            criterion: Some(&["released-timestamp:desc"]),
-            ..BASE_CONF
-        },
-        utils::Conf {
-            group_name: "desc",
-            criterion: Some(&["released-timestamp:desc"]),
-            ..BASE_CONF
-        },
-
         /* then we bench the asc and desc criterion on top of the default criterion */
         utils::Conf {
             group_name: "asc + default",
             criterion: Some(&asc_default[..]),
             ..BASE_CONF
         },
-        utils::Conf {
-            group_name: "desc + default",
-            criterion: Some(&desc_default[..]),
-            ..BASE_CONF
-        },
-
-        /* we bench the filters with the default request */
-        utils::Conf {
-            group_name: "basic filter: <=",
-            filter: Some("released-timestamp <= 946728000"), // year 2000
-            ..BASE_CONF
-        },
-        utils::Conf {
-            group_name: "basic filter: TO",
-            filter: Some("released-timestamp 946728000 TO 1262347200"), // year 2000 to 2010
-            ..BASE_CONF
-        },
-        utils::Conf {
-            group_name: "big filter",
-            filter: Some("released-timestamp != 1262347200 AND (NOT (released-timestamp = 946728000)) AND (duration-float = 1 OR (duration-float 1.1 TO 1.5 AND released-timestamp > 315576000))"),
-            ..BASE_CONF
-        },
-
-        /* the we bench some global / normal search with all the default criterion in the default
-         * order */
-        utils::Conf {
-            group_name: "basic placeholder",
-            queries: &[""],
-            ..BASE_CONF
-        },
-        utils::Conf {
-            group_name: "basic without quote",
-            queries: &BASE_CONF
-                .queries
-                .iter()
-                .map(|s| s.trim()) // we remove the space at the end of each request
-                .collect::<Vec<&str>>(),
-            ..BASE_CONF
-        },
-        utils::Conf {
-            group_name: "basic with quote",
-            queries: basic_with_quote,
-            ..BASE_CONF
-        },
-        utils::Conf {
-            group_name: "prefix search",
-            queries: &[
-                "s", // 500k+ results
-                "a", //
-                "b", //
-                "i", //
-                "x", // only 7k results
-            ],
-            ..BASE_CONF
-        },
     ];
 
     utils::run_benches(c, confs);
diff --git a/crates/benchmarks/benches/sort.rs b/crates/benchmarks/benches/sort.rs
index c3e934432..c0677eef4 100644
--- a/crates/benchmarks/benches/sort.rs
+++ b/crates/benchmarks/benches/sort.rs
@@ -49,62 +49,11 @@ const BASE_CONF: Conf = Conf {
 fn bench_sort(c: &mut criterion::Criterion) {
     #[rustfmt::skip]
     let confs = &[
-        utils::Conf {
-            group_name: "without sort",
-            sort: None,
-            ..BASE_CONF
-        },
-
-        utils::Conf {
-            group_name: "sort on many different values",
-            sort: Some(vec!["name:asc"]),
-            ..BASE_CONF
-        },
-
-        utils::Conf {
-            group_name: "sort on many similar values",
-            sort: Some(vec!["timezone:desc"]),
-            ..BASE_CONF
-        },
-
-        utils::Conf {
-            group_name: "sort on many similar then different values",
-            sort: Some(vec!["timezone:desc", "name:asc"]),
-            ..BASE_CONF
-        },
-
         utils::Conf {
             group_name: "sort on many different then similar values",
             sort: Some(vec!["timezone:desc", "name:asc"]),
             ..BASE_CONF
         },
-
-        utils::Conf {
-            group_name: "geo sort",
-            sample_size: Some(10),
-            sort: Some(vec!["_geoPoint(45.4777599, 9.1967508):asc"]),
-            ..BASE_CONF
-        },
-
-        utils::Conf {
-            group_name: "sort on many similar values then geo sort",
-            sample_size: Some(50),
-            sort: Some(vec!["timezone:desc", "_geoPoint(45.4777599, 9.1967508):asc"]),
-            ..BASE_CONF
-        },
-
-        utils::Conf {
-            group_name: "sort on many different values then geo sort",
-            sample_size: Some(50),
-            sort: Some(vec!["name:desc", "_geoPoint(45.4777599, 9.1967508):asc"]),
-            ..BASE_CONF
-        },
-
-        utils::Conf {
-            group_name: "sort on many fields",
-            sort: Some(vec!["population:asc", "name:asc", "elevation:asc", "timezone:asc"]),
-            ..BASE_CONF
-        },
     ];
 
     utils::run_benches(c, confs);
diff --git a/crates/milli/src/update/upgrade/mod.rs b/crates/milli/src/update/upgrade/mod.rs
index 01ad677c7..b72efb2e6 100644
--- a/crates/milli/src/update/upgrade/mod.rs
+++ b/crates/milli/src/update/upgrade/mod.rs
@@ -77,9 +77,9 @@ const fn start(from: (u32, u32, u32)) -> Option<usize> {
 ///
 /// This serves as a reminder to consider the proper dumpless upgrade implementation when changing the package version.
 const _CHECK_PACKAGE_CAN_UPGRADE: () = {
-    if start((VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH)).is_none() {
-        panic!("cannot upgrade from latest package version")
-    }
+//    if start((VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH)).is_none() {
+//        panic!("cannot upgrade from latest package version")
+//    }
 };
 
 /// Return true if the cached stats of the index must be regenerated
